Terminate handler

mantaionut opened the pull request:
Fixes #50051.
This PR is based on #50320 and I address the last feedback.
On Windows it is enabled by default. Can be enabled or disabled via USE_CUSTOM_TERMINATE env variable.

This PR adds support for overriding the terminate handler in order to log uncaught exceptions in the threads.
If an exception is thrown and not caught, it will print <Unhandled exception caught in c10/util/AbortHandler.h>
The point of doing this is that in issue #50051, exceptions were thrown but not logged. With this logging system it will be easier to debug it in the future.

malfet commented:
I'm not 100% sure how change proposed here fixes issue it references, or what UX problem it addresses. Can you please write a more detailed description of what problems it helps resolve?

mantaionut commented:
@malfet I have updated the description. Let me know if something is not clear.

iremyux commented:
@malfet @albanD @soulitzer 

Can you please check if this can be approved?

mantaionut commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,82 @@
+ public:
+  static AbortHandlerHelper& getInstance() {
+#ifdef _WIN32
+    thread_local
```
It is global on other platforms but not on windows. https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/set-terminate-crt?view=msvc-170 In a multithreaded environment, terminate functions are maintained separately for each thread. Each new thread needs to install its own terminate function. Thus, each thread is in charge of its own termination handling.

albanD replied:
Ok that is unfortunate... Thanks for the doc!

mantaionut replied:
I tested with setunhandledexceptionfilter and does not work with all exceptions. On further testing, I think the best approach is https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-addvectoredexceptionhandler which catches all exceptions from all threads. However, one drawback of this approach is that if an uncaught exception is happening in other python module we will also catch it. What do you think?

albanD commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,82 @@
+} // namespace detail
+
+C10_ALWAYS_INLINE void set_terminate_handler() {
+  bool use_custom_terminate = false;
```
Can you leave an explicit comment saying that the env variable default is different on windows and why?

mantaionut replied:
I will add a comment. The reason for this is based on this comment on the PR i based this one:https://github.com/pytorch/pytorch/pull/50320#issuecomment-763147062

albanD commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,82 @@
+#endif // _WIN32
+  const char* value_str = std::getenv("USE_CUSTOM_TERMINATE");
+  std::string value{value_str != nullptr ? value_str : ""};
+  if (!value.empty()) {
```
You can use `c10::utils::check_env()` to parse the env variable

albanD commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,82 @@
+  std::cout << "Unhandled exception caught in c10/util/AbortHandler.h"
+            << std::endl;
+  auto backtrace = get_backtrace();
+  std::cout << backtrace << std::endl;
```
Should we do some flushing here to ensure everything is processed before aborting?

malfet commented on torch/csrc/Module.cpp:
```diff
@@ -216,6 +217,14 @@ static PyObject* THPModule_crashIfATenASAN(PyObject* module, PyObject* arg) {
   return THPUtils_packInt32(at::_crash_if_asan(THPUtils_unpackInt(arg)));
 }
 
+static PyObject* THPModule_crashImmediately(
+    PyObject* module,
+    PyObject* noargs) {
+  throw std::runtime_error(
+      "Emulate uncaught exception that leads to the crash");
```
It feels wrong to raise uncaptured exception from Python function, as API explicitly states that this is wrong. If you want to abort program execution, why not use `abort` or just raise a signal?
```suggestion
static PyObject* THPModule_abort(
    PyObject* module,
    PyObject* noargs) {
    abort();
```

malfet commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,82 @@
+    }
+  }
+
+  std::terminate_handler getPrev() {
```
Nit (get methods should be marked const)
```suggestion
  std::terminate_handler getPrev() const {
```

mantaionut commented:
@malfet @albanD I addressed the feedback. Regarding using https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-setunhandledexceptionfilter, it only handles low-level exceptions so it's not the best option in our case. With https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-addvectoredexceptionhandler I did not have success since this catches every exception even if there is a try catch. So I went with std::terminate since it behaves the same way as on Linux, except that it only works per thread.

albanD commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,81 @@
+  auto result = c10::utils::check_env("USE_CUSTOM_TERMINATE");
+  if (result != c10::nullopt) {
+    use_custom_terminate = result.value();
+    std::cout << use_custom_terminate << "\n";
```
debug print?

mantaionut replied:
My bad.

albanD commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,81 @@
+#ifdef _WIN32
+  use_custom_terminate = true;
+#endif // _WIN32
+  auto result = c10::utils::check_env("USE_CUSTOM_TERMINATE");
```
Please use TORCH in the name. Something like `TORCH_CUSTOM_TERMINATE=1`

mantaionut replied:
Done

albanD reviewed (commented):
only nits on the impl. I'll let @malfet give his opinion on the particular handler to use and if it is ok.

mantaionut committed:
Add custom Terminate handler

On Windows this handler is enabled by default.
Can be enabled or disabled by using env TORCH_CUSTOM_TERMINATE

albanD reviewed (approved):
Sounds good to me. Thanks for the update.
Let's see if @malfet has any last comment before merging.

kalaskarsanket commented:
@malfet  following up on this pull request and if it is good to merge ?

malfet commented on c10/util/AbortHandler.h:
```diff
@@ -0,0 +1,80 @@
+
+  void set(std::terminate_handler handler) {
+    std::lock_guard<std::mutex> lk(mutex);
+    if (!inited) {
```
Just curious, shouldn't one raise an exception or something if `set` is called twice?

mantaionut replied:
I don't think so, since if someone calls it twice it will do nothing and it is called multiple times anyway since on Windows, it must be called per thread.

mantaionut commented:
@pytorchbot merge

mantaionut commented:
@pytorchbot merge
